src/main/java/org/citopt/connde/service/testing/TestEngine.java

```diff
@@ -0,0 +1,440 @@
+        for (int i = 0; i < test.getSensor().size(); i++) {
+            for (RuleTrigger trigger : allRules) {
+                Sensor sensor = test.getSensor().get(i);
+                String s = sensor.getId();
+                if (trigger.getQuery().contains(s)) {
+                    for (Rule nextRule : ruleRepository.findAll()) {
+                        if (nextRule.getTrigger().getId().equals(trigger.getId())) {
+                            if (!rulesbefore.contains(nextRule)) {
+                                rulesbefore.add(nextRule);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        return rulesbefore;
+    }
+
+    /**
+     * Starts the test and saves all values form the sensor.
+     *
+     * @param test test to be executed
+     */
+    public Map<String, List<Double>> executeTest(TestDetails test) {
+
+        Map<String, TestDetails> activeTests = testEngine.getActiveTests();
+        Map<String, List<Double>> list = testEngine.getTestValues();
+        for (Sensor sensor : test.getSensor()) {
+            activeTests.put(sensor.getId(), test);
+            list.remove(sensor.getId());
+        }
+        testEngine.setActiveTests(activeTests);
+        testEngine.setTestValues(list);
+        testEngine.startTest(testDetailsRepository.findById(test.getId()));
+
+        Map<String, List<Double>> valueList;
+        Map<String, List<Double>> valueListTest = new HashMap<>();
+
+        // Get List of all simulated Values
+        valueList = testEngine.isFinished(test.getId());
+        TestDetails testDetails2 = testDetailsRepository.findOne(test.getId());
+        for (Sensor sensor : test.getSensor()) {
+            List<Double> temp = valueList.get(sensor.getId());
+            valueList.put(sensor.getName(), temp);
+            valueListTest.put(sensor.getName(), temp);
+            list.remove(sensor.getId());
+        }
+
+        // save list of sensor values to database
+        testDetails2.setSimulationList(valueListTest);
+        testDetailsRepository.save(testDetails2);
+
+        return valueListTest;
+    }
+
+
+    /**
+     * Gets all Rules executed by the test
+     *
+     * @param triggerValues map of all trigger values of a specific test
+     * @return a list of all executed rules
+     */
+    public List<String> getRulesExecuted(Map<String, List<Double>> triggerValues) {
+
+        List<String> executedRules = new ArrayList<>();
+        triggerValues.forEach((k, v) -> executedRules.add(k));
+
+        return executedRules;
+    }
+
+    /**
+     * Opens the Test-Report
+     *
+     * @param testId ID of the specific test
+     */
+    public ResponseEntity<String> openPDF(String testId) {
+
+        ResponseEntity<String> response;
+        try {
+            TestDetails testDetails = testDetailsRepository.findById(testId);
+            Desktop.getDesktop().open(new File(testDetails.getPathPDF()));
```

Comment:
Seems like this opens the PDF file on the desktop of the server? But from the perspective of the user this only works if the user is running the MBP on his machine himself, doesn't it? I think this should be removed or replaced by a download that is offered to the user.